### PR TITLE
Update views

### PR DESCRIPTION
**Pull request by dwillist**
## Summary
The goal of this PR is to make the UI components that dive uses re-usable and extendable. This is a requirement to eventually allow users the ability to plugin their own functionality. 

 It introduces a few major changes to do this:
- migrate off of `gocui` to a `tcell` based alternative `tview`, this project gives us some easily extendable types and is actively maintained. 
- refactor front end to have a clean division between views and the data they consume. 

## TODO
That said there are still quite a few features that need to be added before this can be considered equivalent to the previous implementation, namely:
#### Visual changes needed:
  - ~~Pull color settings from terminal / allow configuration~~
  - Could probably use a bit more beautification on the front end.
  - ~~Key mapping view~~
  - ~~Hide filter view when empty/unfocused~~
  - ~~Limited space views~~
#### Functionality features needed:
  - ~~missing some change based filtering (▏^R Removed ▏^M Modified ▏^U Unmodified ▏^B Attributes )~~
  - ~~Few filetree commands missing (collapse all)~~
  - ~~`CompareAllLayers` view~~
  - Factor out the `tview.Grid` used to arrange view components. 

Would love some feedback on the direction of this PR. And will be adding some more documentation to the code to help with readability.

### CODE REVIEW

**Comment by dwillist** [comment]
Hey, @wagoodman when you get the chance, I would love some feedback on the direction of the PR.

( PS, the final commit 38ccd19 is just showing how this could easily be extended. )

**Comment by wagoodman** [comment]
@dwillist this looks like a great! I took a look over the weekend and started to get a feel for the changes and approaches. I have some comments that are probably best paired with some commits, mind if I push to this branch?

**Comment by wagoodman** [comment]
@dwillist I got some time this morning and [pushed a commit](https://github.com/wagoodman/dive/pull/326/commits/4c5949413107ec8d295f91bcd6390a1c0ef4a64e) on what I was working on, specifically to augment the primitives to have a wrapper for consistent styling. I haven't accounted for the layerDetails view yet (and a few more things), but I wanted to start augmenting the foundation to allow for a consistent look-and feel (based on what's there today).

Next I'll take a look at the CNB specifics and the abstractions there.

**Comment by dwillist** [comment]
@wagoodman awesome! Really like the `wrapper` primitive you defined that was a great idea.

I've just added a couple more components to get some "hiding" behavior to help de-clutter views when using a smaller window. Think I'll reimplement the coloring from the terminal next and really I need to add some comments to these primitives & models to make this a bit more readable.

File: runtime/ui/app.go [hunk 1]
```diff
@@ -1,149 +1,159 @@
...
-}
-
 // Run is the UI entrypoint.
-func Run(analysis *image.AnalysisResult, treeStack filetree.Comparer) error {
-	var err error
-
-	g, err := gocui.NewGui(gocui.OutputNormal)
+func Run(analysis *image.AnalysisResult, treeStack filetree.Comparer, isCNB bool) error {
+	cfg := zap.NewDevelopmentConfig()
+	cfg.OutputPaths = []string{"/tmp/dive/debug.out"}
```
Comment (cxsu) [inline runtime/ui/app.go:136]:
```suggestion
	cfg := zap.NewDevelopmentConfig()
	os.MkdirAll("/tmp/dive", os.ModePerm)
	cfg.OutputPaths = []string{"/tmp/dive/debug.out"}
```

I added it to correct the error below.

> panic: couldn't open sink ... no such file or directory

Comment (dwillist) [reply runtime/ui/app.go:136]: :+1:  much appreciated

**Comment by dwillist** [comment]
Hey sorry this has taken so long but were getting close! There are still a couple features that need to be added (listed below). But in general this is quite close to being finished. It would be tremendously helpful if I could get a review started to collect feedback while wrapping things up as this is a pretty substantial change.

- comments & docs !!!!! :( 
- recently added `wrap` option
- a compatibility layer between old `dive.config`, and currently implemented one
- some of the old settings are not yet incorporated before app initialization
  - collapsing all folders
  - setting filetree width
  - hiding some filetypes on startup (added, removed, ....) 
 

Some future TODO's that I would consider out of scope in the PR but want to add later:
- CNB behavior
- Initialization helpers to reduce complexity in `runtime/ui/app.go`. These would also likely be where we could start considering what using `gRPC` looks like.
- Mouse? resizable windows (not too much work anymore thanks `tview` )
- Details scrolling


PS all of the `CNB` specific behavior has been removed and will be added back in later.

**Comment by dwillist** [comment]
@wagoodman hey wondering if you're still interested in this PR or have gotten a chance to look at it

**Comment by wagoodman** [comment]
@dwillist ... just getting back to this since the new years break --this is my first week back from paternity leave, it's been a blast! Thanks for your work on this --this is awesome. I'll be digging into the code and details today and adding comments and pushing tweaks.

**Comment by wagoodman** [comment]
@dwillist I'm currently working on changes to your branch that would firm up how logging and application configuration is initialized and handled within dive. I think this would help the foundation some and what is needed for `ui.UI` initialization --shout out if you're working in these areas and we'll coordinate.

**Comment by wagoodman** [comment]
@dwillist  I updated the app config logic and logging facilities (there was apex/log, logrus, and zap in use). I saw that you had some work that was still in progress and not pushed yet in `app.go` and `key_config.go` (grep for `partially pushed`), I commented it out in the meantime to get the branch running.

I'm currently troubleshooting a UI state bug where user options are not persisted when the layer selection is changed.

**Comment by dwillist** [comment]
Hey :tada:  Congrats!! :tada:. Glad to have you back though. Apologies about that have baked commit. I had some changes in flight but moving them over should be ez pz and shouldn't interfere. Also keybinding and log changes look loads better than what I hacked together :+1:

**Comment by wagoodman** [comment]
@dwillist thanks! No worries about the commit 👍  (I'm dipping my toes on your branch, so it's me getting in the way :) ). I've got some extra time today, I'm going to take a stab at continuing the keybinding config (great find on `cbind` btw). Next on the list for me is to take a closer look at the models interface approach.

**Comment by abitrolly** [comment]
Wow, lots of conflicts here. Seems useful.

**Pull request closed without merging**